private/pkg/wasm/runtime.go

```diff
@@ -0,0 +1,131 @@
+// Copyright 2020-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package wasm
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/bufbuild/buf/private/pkg/syserror"
+	"github.com/tetratelabs/wazero"
+	"github.com/tetratelabs/wazero/api"
+	"github.com/tetratelabs/wazero/imports/wasi_snapshot_preview1"
+	"go.uber.org/multierr"
+)
+
+const (
+	// defaultMaxMemoryBytes is the maximum memory size in bytes.
+	defaultMaxMemoryBytes = 1 << 29 // 512 MiB
+	// wasmPageSize is the page size in bytes.
+	wasmPageSize = 1 << 16 // 64 KiB
+)
+
+type runtime struct {
+	runtime wazero.Runtime
+	cache   wazero.CompilationCache
+}
+
+var _ Runtime = (*runtime)(nil)
+
+func newRuntime(ctx context.Context, options ...RuntimeOption) (*runtime, error) {
+	var cfg runtimeConfig
+	for _, opt := range options {
+		opt.apply(&cfg)
+	}
+	// Create the runtime config with enforceable limits.
+	runtimeConfig := wazero.NewRuntimeConfig().
+		WithCoreFeatures(api.CoreFeaturesV2).
+		WithCloseOnContextDone(true).
+		WithMemoryLimitPages(cfg.getMaxMemoryBytes() / wasmPageSize)
+	var cache wazero.CompilationCache
+	if cfg.cacheDir != "" {
+		var err error
+		cache, err = wazero.NewCompilationCacheWithDir(cfg.cacheDir)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create compilation cache: %w", err)
+		}
+		runtimeConfig = runtimeConfig.WithCompilationCache(cache)
+	}
+	r := wazero.NewRuntimeWithConfig(ctx, runtimeConfig)
```

Comment:
We don't use single character variable names in this codebase beyond method receivers